File: tests/qos/test_robot_controller.cpp
// test_robot_controller.cpp — Unit tests for the robot controller state
// machine. Verifies state transitions, safety interlock behavior, and input
// processing without any DDS dependency.

#include <gtest/gtest.h>
#include "robot_controller.hpp"

using namespace medtech::surgical;

class RobotControllerTest : public ::testing::Test {
protected:
    void SetUp() override
    {
        controller_ = std::make_unique<RobotController>("robot-test-001");
    }

    std::unique_ptr<RobotController> controller_;
};

TEST_F(RobotControllerTest, InitialState)
{
    auto snap = controller_->snapshot();
    EXPECT_EQ(snap.state.robot_id, "robot-test-001");
    EXPECT_EQ(snap.state.operational_mode, Surgery::RobotMode::IDLE);
    EXPECT_EQ(snap.state.error_state, 0);
    EXPECT_EQ(snap.state.joint_positions.size(), 7u);
    EXPECT_FALSE(snap.interlock_active);
}

TEST_F(RobotControllerTest, SafetyInterlockHaltsRobot)
{
    // First make it operational via a command
    Surgery::RobotCommand cmd;
    cmd.robot_id = "robot-test-001";
    cmd.command_id = 1;
    cmd.target_position.x = 1.0;
    cmd.target_position.y = 2.0;
    cmd.target_position.z = 3.0;
    controller_->apply_robot_command(cmd);

    auto snap = controller_->snapshot();
    EXPECT_EQ(snap.state.operational_mode, Surgery::RobotMode::OPERATIONAL);

    // Activate interlock
    Surgery::SafetyInterlock interlock;
    interlock.robot_id = "robot-test-001";
    interlock.interlock_active = true;
    interlock.reason = "Test interlock";
    controller_->apply_safety_interlock(interlock);

    snap = controller_->snapshot();
    EXPECT_EQ(snap.state.operational_mode, Surgery::RobotMode::EMERGENCY_STOP);
    EXPECT_TRUE(snap.interlock_active);
    EXPECT_EQ(snap.state.error_state, 1);
}

TEST_F(RobotControllerTest, InterlockClearResumesToIdle)
{
    // Activate interlock
    Surgery::SafetyInterlock interlock;
    interlock.robot_id = "robot-test-001";
    interlock.interlock_active = true;
    interlock.reason = "Test";
    controller_->apply_safety_interlock(interlock);

    EXPECT_EQ(controller_->snapshot().state.operational_mode,
              Surgery::RobotMode::EMERGENCY_STOP);

    // Clear interlock
    interlock.interlock_active = false;
    controller_->apply_safety_interlock(interlock);

    auto snap = controller_->snapshot();
    EXPECT_EQ(snap.state.operational_mode, Surgery::RobotMode::IDLE);
    EXPECT_FALSE(snap.interlock_active);
    EXPECT_EQ(snap.state.error_state, 0);
}

TEST_F(RobotControllerTest, OperatorInputDiscardedDuringInterlock)
{
    // Activate interlock
    Surgery::SafetyInterlock interlock;
    interlock.robot_id = "robot-test-001";
    interlock.interlock_active = true;
    interlock.reason = "Blocked";
    controller_->apply_safety_interlock(interlock);

    // Joint positions before input
    auto before = controller_->snapshot().state.joint_positions;

    // Send operator input — should be discarded
    Surgery::OperatorInput input;
    input.operator_id = "op-001";
    input.robot_id = "robot-test-001";
    input.x_axis = 100.0;
    input.y_axis = 100.0;
    input.z_axis = 100.0;
    controller_->apply_operator_input(input);

    // Verify joints did not change
    auto after = controller_->snapshot().state.joint_positions;
    EXPECT_EQ(before, after);
    EXPECT_EQ(controller_->snapshot().state.operational_mode,
              Surgery::RobotMode::EMERGENCY_STOP);
}

TEST_F(RobotControllerTest, RobotCommandAppliesTargetPosition)
{
    Surgery::RobotCommand cmd;
    cmd.robot_id = "robot-test-001";
    cmd.command_id = 42;
    cmd.target_position.x = 1.5;
    cmd.target_position.y = 2.5;
    cmd.target_position.z = 3.5;
    controller_->apply_robot_command(cmd);

    auto snap = controller_->snapshot();
    EXPECT_EQ(snap.state.operational_mode, Surgery::RobotMode::OPERATIONAL);
    EXPECT_DOUBLE_EQ(snap.state.tool_tip_position.x, 1.5);
    EXPECT_DOUBLE_EQ(snap.state.tool_tip_position.y, 2.5);
    EXPECT_DOUBLE_EQ(snap.state.tool_tip_position.z, 3.5);
}

TEST_F(RobotControllerTest, RobotCommandIgnoredDuringInterlock)
{
    // Activate interlock
    Surgery::SafetyInterlock interlock;
    interlock.robot_id = "robot-test-001";
    interlock.interlock_active = true;
    interlock.reason = "Blocked";
    controller_->apply_safety_interlock(interlock);

    auto before = controller_->snapshot().state.tool_tip_position;

    // Send command — should be ignored
    Surgery::RobotCommand cmd;
    cmd.robot_id = "robot-test-001";
    cmd.command_id = 1;
    cmd.target_position.x = 99.0;
    controller_->apply_robot_command(cmd);

    auto after = controller_->snapshot().state.tool_tip_position;
    EXPECT_DOUBLE_EQ(before.x, after.x);
}

TEST_F(RobotControllerTest, OperatorInputMovesJoints)
{
    Surgery::OperatorInput input;
    input.operator_id = "op-001";
    input.robot_id = "robot-test-001";
    input.x_axis = 1.0;
    input.y_axis = 2.0;
    input.z_axis = 3.0;
    input.roll = 0.0;
    input.pitch = 0.0;
    input.yaw = 0.0;
    controller_->apply_operator_input(input);

    auto snap = controller_->snapshot();
    // Input transitions to OPERATIONAL
    EXPECT_EQ(snap.state.operational_mode, Surgery::RobotMode::OPERATIONAL);
    // First 3 joints should have changed
    EXPECT_NE(snap.state.joint_positions[0], 0.0);
    EXPECT_NE(snap.state.joint_positions[1], 0.0);
    EXPECT_NE(snap.state.joint_positions[2], 0.0);
}

TEST_F(RobotControllerTest, ReliableCommandOrdering)
{
    // Send multiple commands — each should update the target position
    for (int i = 1; i <= 5; ++i) {
        Surgery::RobotCommand cmd;
        cmd.robot_id = "robot-test-001";
        cmd.command_id = i;
        cmd.target_position.x = static_cast<double>(i);
        cmd.target_position.y = static_cast<double>(i * 10);
        cmd.target_position.z = 0.0;
        controller_->apply_robot_command(cmd);
    }

    // Final state should reflect the last command
    auto snap = controller_->snapshot();
    EXPECT_DOUBLE_EQ(snap.state.tool_tip_position.x, 5.0);
    EXPECT_DOUBLE_EQ(snap.state.tool_tip_position.y, 50.0);
}

File: tests/integration/test_robot_controller.py
"""Tests for Step 2.2 — Robot Simulator DDS QoS behaviors.

Spec coverage: surgical-procedure.md — Robot Teleop & Control
Tags: @integration @streaming @command

Tests the DDS QoS behavior for control topics (RobotState, OperatorInput,
RobotCommand, SafetyInterlock) using in-process writer/reader pairs.
The robot controller state machine logic is tested separately via C++ GTest
(test_robot_controller.cpp).
"""

from __future__ import annotations

import time

import pytest
import rti.connextdds as dds
import surgery
from conftest import wait_for_data, wait_for_discovery

pytestmark = [pytest.mark.integration]

# Type aliases
RobotState = surgery.Surgery.RobotState
RobotCommand = surgery.Surgery.RobotCommand
SafetyInterlock = surgery.Surgery.SafetyInterlock
OperatorInput = surgery.Surgery.OperatorInput
RobotMode = surgery.Surgery.RobotMode


class TestRobotStateQoS:
    """Tests for RobotState DDS QoS behavior."""

    def test_robot_state_published_with_correct_fields(
        self, participant_factory, writer_factory, reader_factory
    ):
        """Robot state is published with all required fields.

        Spec: RobotState samples contain joint positions, operational mode,
              and error state.
        """
        partition = "room/OR-1/procedure/rsf-test"
        writer_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        reader_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        provider = dds.QosProvider.default

        topic_w = dds.Topic(writer_p, "RobotState", RobotState)
        topic_r = dds.Topic(reader_p, "RobotState", RobotState)

        writer_qos = provider.datawriter_qos_from_profile("TopicProfiles::RobotState")
        reader_qos = provider.datareader_qos_from_profile("TopicProfiles::RobotState")

        writer = writer_factory(writer_p, topic_w, qos=writer_qos)
        reader = reader_factory(reader_p, topic_r, qos=reader_qos)

        assert wait_for_discovery(writer, reader, timeout_sec=10)

        state = RobotState(
            robot_id="robot-001",
            joint_positions=[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7],
            tool_tip_position=surgery.Surgery.CartesianPosition(x=1.0, y=2.0, z=3.0),
            operational_mode=RobotMode.OPERATIONAL,
            error_state=0,
        )
        writer.write(state)

        received = wait_for_data(reader, timeout_sec=5)
        assert len(received) >= 1
        data = received[0].data
        assert data.robot_id == "robot-001"
        assert len(data.joint_positions) == 7
        assert abs(data.joint_positions[0] - 0.1) < 1e-9
        assert data.operational_mode == RobotMode.OPERATIONAL
        assert data.error_state == 0
        assert abs(data.tool_tip_position.x - 1.0) < 1e-9

    def test_robot_state_late_joiner_receives_state(
        self, participant_factory, writer_factory, reader_factory
    ):
        """Late joiner receives current RobotState via TRANSIENT_LOCAL.

        Spec: RobotState uses State QoS with TRANSIENT_LOCAL durability.
        """
        partition = "room/OR-2/procedure/rslj-test"
        writer_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        provider = dds.QosProvider.default

        topic_w = dds.Topic(writer_p, "RobotState", RobotState)
        writer_qos = provider.datawriter_qos_from_profile("TopicProfiles::RobotState")
        writer = writer_factory(writer_p, topic_w, qos=writer_qos)

        # Publish BEFORE reader exists
        state = RobotState(
            robot_id="robot-late",
            joint_positions=[0.0] * 7,
            operational_mode=RobotMode.IDLE,
            error_state=0,
        )
        writer.write(state)
        time.sleep(0.5)

        # Late-joining reader
        reader_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        topic_r = dds.Topic(reader_p, "RobotState", RobotState)
        reader_qos = provider.datareader_qos_from_profile("TopicProfiles::RobotState")
        reader = reader_factory(reader_p, topic_r, qos=reader_qos)

        received = wait_for_data(reader, timeout_sec=5)
        assert len(received) >= 1
        assert received[0].data.robot_id == "robot-late"


class TestOperatorInputQoS:
    """Tests for OperatorInput DDS QoS behavior."""

    def test_operator_input_delivery(
        self, participant_factory, writer_factory, reader_factory
    ):
        """Operator input is delivered to subscriber.

        Spec: OperatorInput published with TopicProfiles::OperatorInput QoS
              is delivered to subscriber in the same partition.
        """
        partition = "room/OR-3/procedure/oi-test"
        writer_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        reader_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        provider = dds.QosProvider.default

        topic_w = dds.Topic(writer_p, "OperatorInput", OperatorInput)
        topic_r = dds.Topic(reader_p, "OperatorInput", OperatorInput)
        w_qos = provider.datawriter_qos_from_profile("TopicProfiles::OperatorInput")
        r_qos = provider.datareader_qos_from_profile("TopicProfiles::OperatorInput")

        writer = writer_factory(writer_p, topic_w, qos=w_qos)
        reader = reader_factory(reader_p, topic_r, qos=r_qos)

        assert wait_for_discovery(writer, reader, timeout_sec=10)

        op_input = OperatorInput(
            operator_id="op-001",
            robot_id="robot-001",
            x_axis=5.0,
            y_axis=-3.0,
            z_axis=1.0,
            roll=0.5,
            pitch=-0.3,
            yaw=0.1,
        )
        writer.write(op_input)

        # OperatorInput has a 20 ms lifespan — poll with short interval
        deadline = time.time() + 5.0
        received = []
        while time.time() < deadline:
            samples = reader.take()
            valid = [s for s in samples if s.info.valid]
            if valid:
                received = valid
                break
            time.sleep(0.002)  # 2 ms poll — must beat 20 ms lifespan
        assert len(received) >= 1, "No OperatorInput received"
        data = received[0].data
        assert data.operator_id == "op-001"
        assert abs(data.x_axis - 5.0) < 1e-9

    def test_stale_operator_input_discarded_by_lifespan(
        self, participant_factory, writer_factory, reader_factory
    ):
        """Stale operator input (expired lifespan) is not applied.

        Spec: Given OperatorInput with Lifespan20ms snippet (lifespan = 20 ms)
              When the robot controller reads a sample older than 20 ms
              Then the sample is discarded by DDS before delivery.

        Verified by writing a sample, waiting well beyond the lifespan,
        and confirming take() returns no valid samples.
        """
        partition = "room/OR-3/procedure/lifespan-test"
        writer_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        reader_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        provider = dds.QosProvider.default

        topic_w = dds.Topic(writer_p, "OperatorInput", OperatorInput)
        topic_r = dds.Topic(reader_p, "OperatorInput", OperatorInput)
        w_qos = provider.datawriter_qos_from_profile("TopicProfiles::OperatorInput")
        r_qos = provider.datareader_qos_from_profile("TopicProfiles::OperatorInput")

        writer = writer_factory(writer_p, topic_w, qos=w_qos)
        reader = reader_factory(reader_p, topic_r, qos=r_qos)

        assert wait_for_discovery(writer, reader, timeout_sec=10)

        op_input = OperatorInput(
            operator_id="op-stale",
            robot_id="robot-001",
            x_axis=1.0,
        )
        writer.write(op_input)

        # Wait well beyond 20 ms lifespan
        time.sleep(0.2)

        # Any samples should have expired — take() returns nothing valid
        samples = reader.take()
        valid = [s for s in samples if s.info.valid]
        assert (
            len(valid) == 0
        ), f"Expected 0 valid samples after lifespan expiry, got {len(valid)}"


class TestSafetyInterlockQoS:
    """Tests for SafetyInterlock DDS QoS behavior."""

    def test_safety_interlock_delivery_reliable(
        self, participant_factory, writer_factory, reader_factory
    ):
        """Safety interlock uses reliable delivery with State QoS.

        Spec: SafetyInterlock uses State QoS (RELIABLE, TRANSIENT_LOCAL).
        """
        partition = "room/OR-4/procedure/si-test"
        writer_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        reader_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        provider = dds.QosProvider.default

        topic_w = dds.Topic(writer_p, "SafetyInterlock", SafetyInterlock)
        topic_r = dds.Topic(reader_p, "SafetyInterlock", SafetyInterlock)
        w_qos = provider.datawriter_qos_from_profile("TopicProfiles::SafetyInterlock")
        r_qos = provider.datareader_qos_from_profile("TopicProfiles::SafetyInterlock")

        writer = writer_factory(writer_p, topic_w, qos=w_qos)
        reader = reader_factory(reader_p, topic_r, qos=r_qos)

        assert wait_for_discovery(writer, reader, timeout_sec=10)

        interlock = SafetyInterlock(
            robot_id="robot-001",
            interlock_active=True,
            reason="E-stop pressed",
        )
        writer.write(interlock)

        received = wait_for_data(reader, timeout_sec=5)
        assert len(received) >= 1
        data = received[0].data
        assert data.interlock_active == True  # noqa: E712 — DDS returns int
        assert data.reason == "E-stop pressed"


class TestRobotCommandQoS:
    """Tests for RobotCommand DDS QoS behavior."""

    def test_robot_command_reliable_ordered(
        self, participant_factory, writer_factory, reader_factory
    ):
        """Robot commands are delivered reliably and in order.

        Spec: RobotCommand with Command QoS (RELIABLE, VOLATILE, KEEP_LAST 1)
              delivers commands in publication order.
        """
        partition = "room/OR-5/procedure/cmd-test"
        writer_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        reader_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        provider = dds.QosProvider.default

        topic_w = dds.Topic(writer_p, "RobotCommand", RobotCommand)
        topic_r = dds.Topic(reader_p, "RobotCommand", RobotCommand)
        w_qos = provider.datawriter_qos_from_profile("TopicProfiles::RobotCommand")
        r_qos = provider.datareader_qos_from_profile("TopicProfiles::RobotCommand")

        writer = writer_factory(writer_p, topic_w, qos=w_qos)
        reader = reader_factory(reader_p, topic_r, qos=r_qos)

        assert wait_for_discovery(writer, reader, timeout_sec=10)

        # Send 3 commands with different keys to avoid KEEP_LAST 1 overlap
        for i in range(3):
            cmd = RobotCommand(
                robot_id=f"robot-{i:03d}",
                command_id=i + 1,
                target_position=surgery.Surgery.CartesianPosition(
                    x=float(i), y=float(i * 10), z=0.0
                ),
            )
            writer.write(cmd)

        time.sleep(0.5)
        samples = reader.take()
        valid = [s for s in samples if s.info.valid]

        # Should receive all 3 (different keys = distinct instances)
        assert len(valid) >= 3, f"Expected 3 commands, got {len(valid)}"
        # Verify ordering by checking command_ids are ascending
        ids = [s.data.command_id for s in valid]
        assert ids == sorted(ids), f"Commands out of order: {ids}"

    def test_operator_input_deadline_qos_enforced(
        self, participant_factory, writer_factory, reader_factory
    ):
        """OperatorInput deadline QoS is enforced by DDS.

        Spec: 4 ms deadline enforced by DDS Deadline QoS —
              REQUESTED_DEADLINE_MISSED status on the reader indicates
              a stream interruption.

        We verify the deadline QoS is configured by checking that when
        no OperatorInput arrives for longer than the deadline period,
        the reader's REQUESTED_DEADLINE_MISSED status count increments.
        """
        partition = "room/OR-6/procedure/deadline-test"
        writer_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        reader_p = participant_factory(
            domain_id=0, domain_tag="control", partition=partition
        )
        provider = dds.QosProvider.default

        topic_w = dds.Topic(writer_p, "OperatorInput", OperatorInput)
        topic_r = dds.Topic(reader_p, "OperatorInput", OperatorInput)
        w_qos = provider.datawriter_qos_from_profile("TopicProfiles::OperatorInput")
        r_qos = provider.datareader_qos_from_profile("TopicProfiles::OperatorInput")

        writer = writer_factory(writer_p, topic_w, qos=w_qos)
        reader = reader_factory(reader_p, topic_r, qos=r_qos)

        assert wait_for_discovery(writer, reader, timeout_sec=10)

        # Write one sample to start the deadline clock
        op_input = OperatorInput(
            operator_id="op-deadline",
            robot_id="robot-001",
            x_axis=1.0,
        )
        writer.write(op_input)

        # Wait well past the 4 ms deadline without sending more
        time.sleep(0.5)

        status = reader.requested_deadline_missed_status
        assert status.total_count >= 1, (
            f"Expected deadline miss after 500 ms silence, "
            f"got total_count={status.total_count}"
        )
